src/main/java/org/reaktivity/specification/socks/internal/SocksFunctions.java

```diff
@@ -126,6 +225,62 @@ public String getPrefixName()
         }
     }
 
+    public static void fillInBytes(byte[] addressBytes, int index, String ipv6Group)
+    {
+        int ipv6GroupLength = ipv6Group.length();
+        if  (ipv6GroupLength < 3)
+        {
+            addressBytes[2 * index] = 0;
+            addressBytes[2 * index + 1] = (byte) Integer.parseInt(ipv6Group, 16);
+        }
+        else
+        {
+            int mid = ipv6GroupLength/2;
+            addressBytes[2 * index] = (byte) Integer.parseInt(ipv6Group.substring(0, mid), 16);
+            addressBytes[2 * index + 1] =
+                (byte) Integer.parseInt(ipv6Group.substring(mid, ipv6GroupLength), 16);
+        }
+    }
+
+    public static void fillInIpv6HexCompressed(String address, byte[] addressBytes)
+    {
+        final Matcher ipv6Matcher = IPV6_HEX_COMPRESSED_MATCHER.get().reset(address);
+        ipv6Matcher.matches();
+        int startIndex = 0;
+        int endIndex = 7;
+        for (int i = 0; i < 7; i++)
+        {
+            String ipv6Group = ipv6Matcher.group(i + 1);
+            if (ipv6Group == null)
+            {
+                startIndex = i;
+                break;
+            }
+            else
+            {
+                fillInBytes(addressBytes, i, ipv6Group);
+            }
+        }
+        for (int i = 14; i > 7; i--)
+        {
+            String ipv6Group = ipv6Matcher.group(i + 1);
+            if (ipv6Group == null)
+            {
+                break;
+            }
+            else
+            {
+                fillInBytes(addressBytes, endIndex, ipv6Group);
+                endIndex--;
+            }
+        }
+        for (int i = startIndex; i < endIndex + 1; i++)
+        {
+            addressBytes[2*i] = 0;
+            addressBytes[2*i + 1] = 0;
+        }
```

Comment:
Fill the bytes with zeroes before you start, use `Arrays.fill(..)`
Loop through the groups from start to finish, stop at `null` group.
Loop through the groups from finish to start, stop at `null` group.

Reply:
done